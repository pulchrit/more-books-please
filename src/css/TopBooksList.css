.top-books {
    display: flex;
    flex-direction: column;
    justify-content: center; 
    align-items: space-between;
    padding: 20px;

}

.placeholder-text {
    font-weight: bold;
    color: #30490A;
    margin: 20px;
}

.error {
    color: red;
    margin: 20px;
}

@media screen and (min-width: 481px) {
    .top-books {
        display: grid;
        grid-template-columns: repeat(2, 1fr);
        grid-gap: 20px;
        grid-auto-rows: minmax(50px, auto);
        align-items: start;
        margin-left: auto;
        margin-right: auto;  
    }

    .subhead-top-books {
        grid-column: 1/3;
        margin-bottom: 0;
    }
}

@media screen and (min-width: 601px) {
    .top-books {
        /* display: grid; */
        grid-template-columns: repeat(3, 1fr);
        /* grid-gap: 20px;
        grid-auto-rows: minmax(50px, auto);
        align-items: start;
        margin-left: auto;
        margin-right: auto; */  
    }

    .subhead-top-books {
        grid-column: 1/4;
       /*  margin-bottom: 0; */
    }
}


@media screen and (min-width: 1000px) {
    .top-books {
        /* display: grid; */
        grid-template-columns: repeat(5, 1fr);
        /*grid-gap: 20px;
        grid-auto-rows: minmax(50px, auto);
         align-items: start; */
        max-width: 1200px;
      /*   margin-left: auto;
        margin-right: auto;   */
    }

    .subhead-top-books {
        grid-column: 1/6;
/*         margin-bottom: 0;
 */    }
}