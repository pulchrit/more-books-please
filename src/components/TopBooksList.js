import React from 'react';
import BookItem from './BookItem';
import config from './config';
import { handleErrors, toTitleCase } from './utilities.js'; 
import '../css/TopBooksList.css';

const processTopBooksData = (data) => {
        
    const display_name = data.results.display_name;

    const books = data.results.books.map(book => {
        return {
            rank: book.rank,
            primary_isbn10: book.primary_isbn10,
            description: book.description,
            title: toTitleCase(book.title),
            author: book.author,
            book_image: book.book_image,
            open_library_url: '',
            amazon_product_url: book.amazon_product_url
        }
    });
        
    return {
        display_name, 
        books
    };
}



export default class TopBooksList extends React.Component {
    
    state = {
        topBooksByList: {
            display_name: '',
            books: []
        },
        error: null
    }

    reviseTopBooksByList = (data, topBooksByList) => {
        
        const isbnAndInfoUrl = Object.keys(data).map(key => {
          return {
            isbn: key.slice(5),
            info_url: data[key].info_url
          }
        });
            
        topBooksByList.books.forEach(book => {
            isbnAndInfoUrl.forEach(isbnObject => {
              if (isbnObject.isbn === book.primary_isbn10) {
                book.open_library_url = isbnObject.info_url;
              } 
            });
          });
    
        return topBooksByList;
    }

    // Get OpenLibrary urls, process data, add OpenLibrary urls to 
    // topBooksByList object and setState with this revised topBooksByList object.
    getOpenLibraryUrls = (topBooksByList) => {
        
        const baseOpenLibraryUrl = 'https://openlibrary.org/api/books?bibkeys='
    
        const isbnNumbers = topBooksByList.books.map(book => 
            `ISBN:${book.primary_isbn10}`)
            .join(',');
    
        const callOpenLibraryUrl = `${baseOpenLibraryUrl}${isbnNumbers}&format=json`;
    
        fetch(callOpenLibraryUrl)
            .then(handleErrors)
            .then(response => response.json())   
            .then(data => this.reviseTopBooksByList(data, topBooksByList))
            .then(topBooksByList => this.setState({topBooksByList}))
            .catch(error => this.setState({error: error.message}))
    }

    componentDidMount() {
        
        // Get selected NTY best seller list data, process data into topBooksByList
        // object, and pass that object to getOpenLibraryUrls function to call 
        // the OpenLibrary API (which will then update the topBooksByList object 
        // and setState for topBooksByList).
        const listName = this.props.match.params.listName

        const baseNYTUrl = 'https://api.nytimes.com/svc/books/v3/lists/current/';
    
        const callNYTUrl = `${baseNYTUrl}${listName}?api-key=${config.API_KEY_NYT}`;
      
        fetch(callNYTUrl)
            .then(handleErrors)
            .then(response => response.json())
            .then(data => processTopBooksData(data) || data)
            .then(topBooksByList => this.getOpenLibraryUrls(topBooksByList))
            .catch(error => this.setState({ error }))
    }

    render() {

        const {error} = this.state;

        const bookItems = this.state.topBooksByList.books.map((book) => {
            
            return (
                <BookItem 
                    key={book.rank}
                    rank={book.rank}
                    title={book.title}
                    author={book.author}
                    description={book.description}
                    image={book.book_image}
                    url={book.open_library_url === '' ? book.amazon_product_url : book.open_library_url}
                />
            )}
        );
        
        return (
            <> 
                {/* If there is an error, render it, otherwise 'display' an empty string. */}
                {error ? <p className='error' role='alert'>{error.message}</p> : ''}

                {this.state.topBooksByList.display_name === '' 
                    ? <p className='placeholder-text'>Gathering books...one moment please...</p>
                
                    : <section className="top-books">
                        <h2 className="subhead-top-books">{this.state.topBooksByList.display_name}</h2>
                        {bookItems}
                    </section>}
            </>
        );
    }
}

