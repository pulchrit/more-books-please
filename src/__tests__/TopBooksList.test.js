import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {BrowserRouter, MemoryRouter} from 'react-router-dom';
import TopBooksList from '../components/TopBooksList';

describe ("TopBooksList", () => {

    const topBooksList = {
        display_name: 'Children\'s Chapter Books',
        books: [
            {
                rank: 1,
                primary_isbn10: "0385376715",
                description: "A celebration of future possibilities.",
                title: "THE WONDERFUL THINGS YOU WILL BE",
                author: "Emily Winfield Martin",
                book_image: "https://s1.nyt.com/du/books/images/9780385376716.jpg",
                open_library_url: "https://openlibrary.org/books/OL26629977M/The_Wonderful_Things_You_Will_Be",
                amazon_product_url: "http://www.amazon.com/The-Wonderful-Things-You-Will/dp/0385376715?tag=NYTBS-20"
            }
        ]
    }

    const match = {
            params: {
                listName: "chapter-books"
            }    
    };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <MemoryRouter initialEntries={["/topBooks/chapter-books"]} initialIndex={0}>
            <TopBooksList match={match}/>
        </MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  // This snapshot is passing because it renders the placholder text, and 
  // it does render the placholder text correctly. However, I can't figure 
  // out how to test the updated component after setState is run after the 
  // API calls. I know Enzyme has a setState() method, but I don't believe
  // I can use Enzyme with react-router. 
  it('renders UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/topBooks/chapter-books"]} >
            <TopBooksList 
                match={match}
                topBooksList={topBooksList}
            />
        </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })

})


