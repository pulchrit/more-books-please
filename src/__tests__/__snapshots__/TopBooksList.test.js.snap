// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBooksList renders UI as expected 1`] = `
Array [
  "",
  <p
    className="placeholder-text"
  >
    Gathering books...one moment please...
  </p>,
]
`;
